### PR TITLE
 Move MonoClass:inlinearray_value out of MonoClass

### LINKED ISSUE #103913: [metadata] Move MonoClass:inlinearray_value out of MonoClass
inline arrays are quite infrequent.  We have the `is_inlinearray` bit on MonoClass for fast checks, but `inlinearray_value` doesn't really need to be on every single MonoClass.  We should move it to `MonoClass:infrequent_data`, or at least to `MonoClassDef` (since generic instances can't change a custom attribute value).

### PR DESCRIPTION
**Pull request by syfFerdinand**
# Summary
This PR addresses issue #103913 by moving the inlinearray_value property from MonoClass to the infrequent_data structure. This change aims to optimize memory usage by storing rarely used properties in a less frequently accessed structure.

# Details
- Moved `inlinearray_value` from `MonoClass` to `MonoPropertyBag` within the `infrequent_data` structure.
- Set explicit value for PROP_INLINEARRAY_VALUE in InfrequentDataKind enum, as per review comments.
- Updated relevant getters and setters to reflect the new structure.
- Marked mono_class_get_inlinearray_value with MONO_COMPONENT_API as per review suggestion to ensure proper export when compiled as shared libraries.
- Modified all occurrences of `inlinearray_value` access to use the new structure.
- Performed testing to ensure no functionality is broken due to this change.

# Related Issue
Fixes #103913

### CODE REVIEW

**Comment by syfFerdinand** [comment]
@dotnet-policy-service agree

**Comment by lambdageek** [comment]
Thanks for the contribution @syfFerdinand !

**Review by lambdageek — CHANGES_REQUESTED** [review]

File: src/mono/mono/metadata/property-bag.h [hunk 1]
```diff
@@ -21,6 +22,7 @@ struct _MonoPropertyBagItem {
 
 typedef struct {
 	MonoPropertyBagItem *head;
+	gint32     inlinearray_value;
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply src/mono/mono/metadata/property-bag.h:25]: @syfFerdinand yea https://github.com/syfFerdinand/dotnet-runtime/commit/c18a8cb1c2268697fb37ddba4ab0764f87ed2d4a looks great, thanks!

one minor nit:

when you're adding PROP_INLINEARRAY_VALUE to the enum in class-accessors.c, give it an explicit value (15) like the other values.

@steveisok could someone from your team take care of this PR

[the remark this answers is not in the document]
Comment (syfFerdinand) [reply src/mono/mono/metadata/property-bag.h]: It’s done! I added the explicit value for PROP_INLINEARRAY_VALUE as you suggested. Thanks for the help!

**Comment by steveisok** [comment]
@kg can you please give this a review?

File: src/mono/mono/metadata/class-internals.h [hunk 2]
```diff
@@ -1356,6 +1356,12 @@ mono_class_get_declsec_flags (MonoClass *klass);
 void
 mono_class_set_declsec_flags (MonoClass *klass, guint32 value);
 
+gint32
+mono_class_get_inlinearray_value (MonoClass *klass);
```
Comment (lambdageek) [inline src/mono/mono/metadata/class-internals.h:1360]:
```suggestion
MONO_COMPONENT_API gint32
mono_class_get_inlinearray_value (MonoClass *klass);
```

the code in src/mono/mono/component is sometimes compiled as separate shared libraries. So if they call back into the runtime, those functions need to be marked with `MONO_COMPONENT_API` (in this case, the getter, which is called from the debugger object)

**Review by lambdageek — APPROVED** [review]

File: src/mono/mono/metadata/class-internals.h (hunk 2 above)
Comment (syfFerdinand) [reply src/mono/mono/metadata/class-internals.h]: @lambdageek  done!

**Comment by kg** [comment]
Is it normal for this to change the MonoAOTOffsets? I assume because a field was removed from MonoClass?

**Comment by lambdageek** [comment]
> Is it normal for this to change the MonoAOTOffsets? I assume because a field was removed from MonoClass?

yea that's the most likely reason

**Comment by kg** [comment]
If nobody else gets to it I will try to fix the offsets this week

**Comment by syfFerdinand** [comment]
Hi @kg and @lambdageek,

I'll be working on updating the MonoAOTOffsets. Should I just remove the line DECL_OFFSET2(MonoClass,inlinearray_value,128) from MonoAOTOffsets, or is there something more specific I need to do?

Thanks for your guidance!

**Comment by kg** [comment]
Can you see https://dev.azure.com/dnceng-public/public/_build/results?buildId=865527&view=artifacts&pathAsName=false&type=publishedArtifacts ? I believe there are downloads for the offsets there, which might be what you need.

**Comment by steveisok** [comment]
> Can you see https://dev.azure.com/dnceng-public/public/_build/results?buildId=865527&view=artifacts&pathAsName=false&type=publishedArtifacts ? I believe there are downloads for the offsets there, which might be what you need.

Note https://github.com/dotnet/runtime/pull/109612 added the actual offset files. Any updates should go into these.

**Comment by kg** [comment]
I've never dealt with this problem before, but it appears that if you download each of the offsets artifacts I linked to, it will be a ZIP file with updated versions of the .h files in it. You would then check in these new .h files to update the offsets. I don't think you want to manually edit the offsets.

**Comment by steveisok** [comment]
> I've never dealt with this problem before, but it appears that if you download each of the offsets artifacts I linked to, it will be a ZIP file with updated versions of the .h files in it. You would then check in these new .h files to update the offsets. I don't think you want to manually edit the offsets.

Unless I'm completely mistaken, we don't auto-generate the offsets any longer and they need to be updated in tree. 

@akoeplinger is that correct?

**Comment by akoeplinger** [comment]
We generate them in PR builds so you can just download and check in the updated offsets, like @kg mentioned.

This is the message on the build, let me know if I can make it clearer:
> Detected changes in Mono offsets files. When running in a PR build you can download the MonoAOTOffsets artifacts and update the files in the src/mono/mono/offsets folder.

**Review by steveisok — APPROVED** [review]

**Comment by syfFerdinand** [comment]
Thank you all for your assistance and guidance throughout this process. I have updated the offset files with the latest versions from the build artifacts and committed the changes.

Is there anything else that needs to be corrected or adjusted?

**Comment by syfFerdinand** [comment]
Thank you all for your assistance and guidance throughout this process. I have updated the offset files with the latest versions from the build artifacts and committed the changes.

Is there anything else that needs to be corrected or adjusted?

**Review by kg — APPROVED** [review]

**Comment by syfFerdinand** [comment]
hi @kg, 
I think the tests that are currently failing had been deactivated in this merge... [#108178](https://github.com/dotnet/runtime/pull/108178) so I don't understand why there's still this build error...

**Comment by kg** [comment]
Out of memory issues on the wasm lanes are common, so I don't think the librarytests lane failure indicates a bug you created.

**Comment by kg** [comment]
Thank you for your contribution!

**Pull request merged**